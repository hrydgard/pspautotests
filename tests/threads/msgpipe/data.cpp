#include "shared.h"

struct ReceiveDataThread : public KernelObjectWaitThread {
	ReceiveDataThread(const char *name, SceUID uid, u32 size, int mode = 0, int prio = 0x60)
		: KernelObjectWaitThread(name, uid, 10000, prio), size_(size), mode_(mode) {
		start();
	}

	virtual int wait() {
		char msg[256];
		int received = 0x1337;
		int result = sceKernelReceiveMsgPipe(object_, msg, size_, mode_, &received, &timeout_);
		checkpoint("  ** %s got result: %08x, received = %08x = %.*s", name_, result, received, received, msg);
		return 0;
	}

	u32 size_;
	int mode_;
};

struct SendDataThread : public KernelObjectWaitThread {
	SendDataThread(const char *name, SceUID uid, const char *data, u32 size, int mode = 0, int prio = 0x60)
		: KernelObjectWaitThread(name, uid, 10000, prio), data_(data), size_(size), mode_(mode) {
		start();
	}

	virtual int wait() {
		int sent = 0x1337;
		int result = sceKernelSendMsgPipe(object_, (void *)data_, size_, mode_, &sent, &timeout_);
		checkpoint("  ** %s got result: %08x, sent = %08x", name_, result, sent);
		return 0;
	}

	const char *data_;
	u32 size_;
	int mode_;
};

extern "C" int main(int argc, char *argv[]) {
	checkpointNext("Using a buffer:");
	{
		SceUID msgpipe = sceKernelCreateMsgPipe("msgpipe", PSP_MEMORY_PARTITION_USER, 0, 0x100, NULL);
		ReceiveDataThread wait_r1("receiving thread 1", msgpipe, 4);
		ReceiveDataThread wait_r2("receiving thread 2", msgpipe, 4);
		ReceiveDataThread wait_r3("receiving thread 3", msgpipe, 4);
		SendDataThread wait_s1("sending thread 1", msgpipe, "msg1msg2m", 9);
		SendDataThread wait_s2("sending thread 2", msgpipe, "sg3", 3);
		schedfMsgPipe(msgpipe);
		sceKernelDelayThread(1000);
		sceKernelDeleteMsgPipe(msgpipe);
	}

	checkpointNext("Without a buffer:");
	{
		SceUID msgpipe = sceKernelCreateMsgPipe("msgpipe", PSP_MEMORY_PARTITION_USER, 0, 0, NULL);
		ReceiveDataThread wait_r1("receiving thread 1", msgpipe, 4);
		ReceiveDataThread wait_r2("receiving thread 2", msgpipe, 4);
		ReceiveDataThread wait_r3("receiving thread 3", msgpipe, 4);
		SendDataThread wait_s1("sending thread 1", msgpipe, "msg1msg2m", 9);
		SendDataThread wait_s2("sending thread 2", msgpipe, "sg3", 3);
		schedfMsgPipe(msgpipe);
		sceKernelDelayThread(1000);
		sceKernelDeleteMsgPipe(msgpipe);
	}

	checkpointNext("Using receive priorities:");
	{
		SceUID msgpipe = sceKernelCreateMsgPipe("msgpipe", PSP_MEMORY_PARTITION_USER, 0x1000, 4, NULL);
		ReceiveDataThread wait_r1("receiving thread 1", msgpipe, 4, 0, 0x33);
		ReceiveDataThread wait_r2("receiving thread 2", msgpipe, 4, 0, 0x32);
		ReceiveDataThread wait_r3("receiving thread 3", msgpipe, 4, 0, 0x31);
		SendDataThread wait_s1("sending thread 1", msgpipe, "msg", 3);
		SendDataThread wait_s2("sending thread 2", msgpipe, "1ms", 3);
		SendDataThread wait_s3("sending thread 3", msgpipe, "g2m", 3);
		SendDataThread wait_s4("sending thread 4", msgpipe, "sg3", 3);
		schedfMsgPipe(msgpipe);
		sceKernelDelayThread(1000);
		sceKernelDeleteMsgPipe(msgpipe);
	}

	checkpointNext("Using send priorities:");
	{
		SceUID msgpipe = sceKernelCreateMsgPipe("msgpipe", PSP_MEMORY_PARTITION_USER, 0x100, 4, NULL);
		SendDataThread wait_s1("sending thread 1", msgpipe, "msg", 3, 0, 0x39);
		SendDataThread wait_s2("sending thread 2", msgpipe, "1ms", 3, 0, 0x38);
		SendDataThread wait_s3("sending thread 3", msgpipe, "g2m", 3, 0, 0x37);
		SendDataThread wait_s4("sending thread 4", msgpipe, "sg3", 3, 0, 0x36);
		ReceiveDataThread wait_r1("receiving thread 1", msgpipe, 4, 0, 0x33);
		ReceiveDataThread wait_r2("receiving thread 2", msgpipe, 4, 0, 0x32);
		ReceiveDataThread wait_r3("receiving thread 3", msgpipe, 4, 0, 0x31);
		schedfMsgPipe(msgpipe);
		sceKernelDelayThread(1000);
		sceKernelDeleteMsgPipe(msgpipe);
	}

	checkpointNext("Using ASAP:");
	{
		SceUID msgpipe = sceKernelCreateMsgPipe("msgpipe", PSP_MEMORY_PARTITION_USER, 0, 4, NULL);
		SendDataThread wait_s1("sending thread 1", msgpipe, "msg", 3, 1);
		SendDataThread wait_s2("sending thread 2", msgpipe, "1ms", 3, 1);
		SendDataThread wait_s3("sending thread 3", msgpipe, "g2m", 3, 1);
		SendDataThread wait_s4("sending thread 4", msgpipe, "sg3", 3, 1);
		ReceiveDataThread wait_r1("receiving thread 1", msgpipe, 4, 1);
		ReceiveDataThread wait_r2("receiving thread 2", msgpipe, 4, 1);
		ReceiveDataThread wait_r3("receiving thread 3", msgpipe, 4, 1);
		schedfMsgPipe(msgpipe);
		sceKernelDelayThread(1000);
		sceKernelDeleteMsgPipe(msgpipe);
	}

	checkpointNext("Using ASAP and no buffer:");
	{
		SceUID msgpipe = sceKernelCreateMsgPipe("msgpipe", PSP_MEMORY_PARTITION_USER, 0, 0, NULL);
		SendDataThread wait_s1("sending thread 1", msgpipe, "msg", 3, 1);
		SendDataThread wait_s2("sending thread 2", msgpipe, "1ms", 3, 1);
		SendDataThread wait_s3("sending thread 3", msgpipe, "g2m", 3, 1);
		SendDataThread wait_s4("sending thread 4", msgpipe, "sg3", 3, 1);
		ReceiveDataThread wait_r1("receiving thread 1", msgpipe, 4, 1);
		ReceiveDataThread wait_r2("receiving thread 2", msgpipe, 4, 1);
		ReceiveDataThread wait_r3("receiving thread 3", msgpipe, 4, 1);
		schedfMsgPipe(msgpipe);
		sceKernelDelayThread(1000);
		sceKernelDeleteMsgPipe(msgpipe);
	}

	return 0;
}