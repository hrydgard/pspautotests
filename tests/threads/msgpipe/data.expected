[x] Using a buffer:
[r]   ** started receiving thread 1
[r]   ** started receiving thread 2
[r]   ** started receiving thread 3
[r]   ** sending thread 1 got result: 00000000, sent = 00000009
[r]   ** receiving thread 1 got result: 00000000, received = 00000004 = msg1
[r]   ** receiving thread 2 got result: 00000000, received = 00000004 = msg2
[r]   ** started sending thread 1
[r]   ** sending thread 2 got result: 00000000, sent = 00000003
[r]   ** receiving thread 3 got result: 00000000, received = 00000004 = msg3
[r]   ** started sending thread 2
Msgpipe: OK (size=56, name=msgpipe, attr=00000000, buffer=100, free=100, sending=0, receiving=0)

[x] Without a buffer:
[r]   ** started receiving thread 1
[r]   ** started receiving thread 2
[r]   ** started receiving thread 3
[r]   ** sending thread 1 got result: 00000000, sent = 00000009
[r]   ** receiving thread 1 got result: 00000000, received = 00000004 = msg1
[r]   ** receiving thread 2 got result: 00000000, received = 00000004 = msg2
[r]   ** started sending thread 1
[r]   ** sending thread 2 got result: 00000000, sent = 00000003
[r]   ** receiving thread 3 got result: 00000000, received = 00000004 = msg3
[r]   ** started sending thread 2
Msgpipe: OK (size=56, name=msgpipe, attr=00000000, buffer=0, free=0, sending=0, receiving=0)

[x] Using receive priorities:
[r]   ** started receiving thread 1
[r]   ** started receiving thread 2
[r]   ** started receiving thread 3
[r]   ** sending thread 1 got result: 00000000, sent = 00000003
[r]   ** started sending thread 1
[r]   ** receiving thread 3 got result: 00000000, received = 00000004 = msg1
[r]   ** sending thread 2 got result: 00000000, sent = 00000003
[r]   ** started sending thread 2
[r]   ** receiving thread 2 got result: 00000000, received = 00000004 = msg2
[r]   ** sending thread 3 got result: 00000000, sent = 00000003
[r]   ** started sending thread 3
[r]   ** receiving thread 1 got result: 00000000, received = 00000004 = msg3
[r]   ** sending thread 4 got result: 00000000, sent = 00000003
[r]   ** started sending thread 4
Msgpipe: OK (size=56, name=msgpipe, attr=00001000, buffer=4, free=4, sending=0, receiving=0)

[x] Using send priorities:
[r]   ** sending thread 1 got result: 00000000, sent = 00000003
[r]   ** started sending thread 1
[r]   ** started sending thread 2
[r]   ** started sending thread 3
[r]   ** started sending thread 4
[r]   ** receiving thread 1 got result: 00000000, received = 00000004 = msgs
[r]   ** sending thread 4 got result: 00000000, sent = 00000003
[r]   ** started receiving thread 1
[r]   ** receiving thread 2 got result: 00000000, received = 00000004 = g3g2
[r]   ** sending thread 3 got result: 00000000, sent = 00000003
[r]   ** sending thread 2 got result: 00000000, sent = 00000003
[r]   ** started receiving thread 2
[r]   ** receiving thread 3 got result: 00000000, received = 00000004 = m1ms
[r]   ** started receiving thread 3
Msgpipe: OK (size=56, name=msgpipe, attr=00000100, buffer=4, free=4, sending=0, receiving=0)

[x] Using ASAP:
[r]   ** sending thread 1 got result: 00000000, sent = 00000003
[r]   ** started sending thread 1
[r]   ** sending thread 2 got result: 00000000, sent = 00000001
[r]   ** started sending thread 2
[r]   ** started sending thread 3
[r]   ** started sending thread 4
[r]   ** receiving thread 1 got result: 00000000, received = 00000004 = msg1
[r]   ** sending thread 3 got result: 00000000, sent = 00000003
[r]   ** sending thread 4 got result: 00000000, sent = 00000001
[r]   ** started receiving thread 1
[r]   ** receiving thread 2 got result: 00000000, received = 00000004 = g2ms
[r]   ** started receiving thread 2
[r]   ** started receiving thread 3
Msgpipe: OK (size=56, name=msgpipe, attr=00000000, buffer=4, free=4, sending=0, receiving=1)
[r]   ** receiving thread 3 got result: 800201b5, received = 00000000 = 
[r]   ** stopped receiving thread 3

[x] Using ASAP and no buffer:
[r]   ** started sending thread 1
[r]   ** started sending thread 2
[r]   ** started sending thread 3
[r]   ** started sending thread 4
[r]   ** receiving thread 1 got result: 00000000, received = 00000004 = msg1
[r]   ** sending thread 1 got result: 00000000, sent = 00000003
[r]   ** sending thread 2 got result: 00000000, sent = 00000001
[r]   ** started receiving thread 1
[r]   ** receiving thread 2 got result: 00000000, received = 00000004 = g2ms
[r]   ** sending thread 3 got result: 00000000, sent = 00000003
[r]   ** sending thread 4 got result: 00000000, sent = 00000001
[r]   ** started receiving thread 2
[r]   ** started receiving thread 3
Msgpipe: OK (size=56, name=msgpipe, attr=00000000, buffer=0, free=0, sending=0, receiving=1)
[r]   ** receiving thread 3 got result: 800201b5, received = 00000000 = 
[r]   ** stopped receiving thread 3
